runtime/src/network/tokio.rs

```diff
@@ -0,0 +1,218 @@
+use crate::Error;
+use std::{net::SocketAddr, time::Duration};
+use tokio::{
+    io::{AsyncReadExt as _, AsyncWriteExt as _},
+    net::{
+        tcp::{OwnedReadHalf, OwnedWriteHalf},
+        TcpListener, TcpStream,
+    },
+    time::timeout,
+};
+use tracing::warn;
+
+/// Implementation of [crate::Sink] for the [tokio] runtime.
+pub struct Sink {
+    write_timeout: Duration,
+    sink: OwnedWriteHalf,
+}
+
+impl crate::Sink for Sink {
+    async fn send(&mut self, msg: &[u8]) -> Result<(), Error> {
+        timeout(self.write_timeout, self.sink.write_all(msg))
+            .await
+            .map_err(|_| Error::Timeout)?
+            .map_err(|_| Error::SendFailed)?;
+        Ok(())
+    }
+}
+
+/// Implementation of [crate::Stream] for the [tokio] runtime.
+pub struct Stream {
+    read_timeout: Duration,
+    stream: OwnedReadHalf,
+}
+
+impl crate::Stream for Stream {
+    async fn recv(&mut self, buf: &mut [u8]) -> Result<(), Error> {
+        // Wait for the stream to be readable
+        timeout(self.read_timeout, self.stream.read_exact(buf))
+            .await
+            .map_err(|_| Error::Timeout)?
+            .map_err(|_| Error::RecvFailed)?;
+        Ok(())
+    }
+}
+
+/// Implementation of [crate::Listener] using the [tokio] runtime.
+pub struct Listener {
+    cfg: Config,
+    listener: TcpListener,
+}
+
+impl crate::Listener for Listener {
+    type Sink = Sink;
+    type Stream = Stream;
+
+    async fn accept(&mut self) -> Result<(SocketAddr, Self::Sink, Self::Stream), Error> {
+        // Accept a new TCP stream
+        let (stream, addr) = self.listener.accept().await.map_err(|_| Error::Closed)?;
+
+        // Set TCP_NODELAY if configured
+        if let Some(tcp_nodelay) = self.cfg.tcp_nodelay {
+            if let Err(err) = stream.set_nodelay(tcp_nodelay) {
+                warn!(?err, "failed to set TCP_NODELAY");
+            }
+        }
+
+        // Return the sink and stream
+        let (stream, sink) = stream.into_split();
+        Ok((
+            addr,
+            Sink {
+                write_timeout: self.cfg.write_timeout,
+                sink,
+            },
+            Stream {
+                read_timeout: self.cfg.read_timeout,
+                stream,
+            },
+        ))
+    }
+}
+
+impl axum::serve::Listener for Listener {
```

Comment:
You can remove this. I used to use it for metrics but concluded that "web" traffic should not use this runtime (as it may desire very different connections semantics than the read/write here).